### DISCUSSION TITLE
Question about Vector Insertion and Index Building

### DISCUSSION THREAD

**Discussion by Chisdo** [milvus-io/milvus#5023] [Q&A and General discussion] [answered]
Hi there, firstly, I got a few questions about my experiment with Milvus. Then I do have some questions about vector insertion and when does index build.
--- My experiment setting: ---
    [only start one GPU-enabled Milvus on one computer]
    a. Milvus image: 1.0.0 GPU-milvus with RTX 2080 ti * 1, CPU has 12 core[server level]
    b. metadata: enable with MySQL
    c. Wal: Disable
    d. cache_size: 180 GB
    e. disk is HDD and other settings are default
    f. Then, I create 5 collections and build index at the beggining:
         collection1: L2, 2048-dimension, 1024-index_file_size, IVF_PQ {nlist=30000, m=32}
         collection2~5: L2, 480-dimension, 1024-index_file_size, IVF_PQ {nlist=30000, m=32} * 4
    g. I start 5 threads to insert data to each collection at the same time
    h. Use different single queries to test the collection 1 several times.

--- my questions here ---
i. Does 1. g. [inserting data to the different collection concurrently] would affect insertion speed?

ii. I got extremely poor testing results from 1. h. : the searching time is extremely high around 700 seconds per query. I feel like Milvus did brute-force during the test. While the result from get_collection_stats function shows all segments have already got PQ index yet. So I'm wondering that does my Milvus did use index for searching or just do the brute-force? Besides, does Milvus really build index during inserting? 

iii. I notice that there are two background operations in Milvus: merging segments and build the index. I'm wondering that if both operations are needed, which one would be done first? Also, can I do the merging operation manually?

iv. If I keep inserting continuous data to a partition in a collection with a small auto_flush_interval [1 second]. May the small auto_flush_interval number would increase the number of segments at this partition?

v. In "Performance Tuning" section, there is a sentence says: "In scenario with continuous data insertion, because Milvus does not index segments with a size less than index_file_size, it uses brute-force search as the query method. ". Does it mean that once the segment is inserted over index_file_size, the segment would be built index immediately?

**Comment by shengjun1985** [comment, marked as answer]
> i. Does 1. g. [inserting data to the different collection concurrently] would affect insertion speed?

No, it doesn't.
Disk IO is the main factor that affects the insertion performance. Multi-threading doesn't help it.

> ii. So I'm wondering that does my Milvus did use index for searching or just do the brute-force?
Besides, does Milvus really build index during inserting?

Please refer to [create_index](https://milvus.io/docs/v1.0.0/index.md#Create-indexes).
According to your description, you firstly specified the index and then inserted the data. So the data inserted must reach index_ file_size (1024 Mb), the index would be created. How about your row_count of the collection 1?
But if your data is static, you can insert the data before specifying the index.
Furthermore, how about the 'nq', 'topk', 'nprobe' of your query?

> iii. I'm wondering that if both operations are needed, which one would be done first? 
Also, can I do the merging operation manually?

Please refer to [merge-data](https://milvus.io/docs/v1.0.0/storage_operation.md#Merge-data)
These two operations depend on the parameter index_ file_size. So they don't work on one segment at the same time.
For different data segments, one is disk IO and the other is CPU computing, so they can be done at the same time.
However, there is no interface for manually merging at present.

> iv. If I keep inserting continuous data to a partition in a collection with a small auto_flush_interval [1 second]. May the small auto_flush_interval number would increase the number of segments at this partition?

Yes, it will in a short time.
And after the data merged, there will be no such problem.

> v. Does it mean that once the segment is inserted over index_file_size, the segment would be built index immediately?

Need to wait for merging segments.

**Reply by Chisdo** [reply]
a. supplement to ii: The row_count of collection 1 is around 78M and almost evenly separates into 11 partitions. For the query part, 'nq' is 1 [my understanding of 'nq' is the length of the query vector lists, isn't it? ], 'topK' is 10k, 'nprobe' is 375. 

b. supplement questions to v: 
refer:  [Build-indexes-during-free-time](https://milvus.io/docs/v1.0.0/index.md#Build-indexes-during-free-time)
After merging segments, [without any query command, just keep inserting streaming data], would Milvus start to build index immediately? Will the index-building operation be interrupted by data insertion?

c. If I only have one GPU, can I use this GPU to build all 5 collections at the same time? How about if I have multiple GPUs?

d. refer: [Section-Cache](https://milvus.io/docs/v1.0.0/milvus_config.md#Section-cache)
 - For 'preload_collection', after preloading some collections, would these collections keep in Cache stablely[cannot be replaced]? 
 - If the size of real 'preload_collection' is over than 'cache_size', what will happen?
 - Is there any way for estimating 'preload_collection' size?

**Reply by shengjun1985** [reply]
> a. the result from get_collection_stats function shows all segments have already got PQ index yet.
The row_count of collection 1 is around 78M and almost evenly separates into 11 partitions
'nq' is 1, 'topK' is 10k, 'nprobe' is 375.

There may be a bug in 'get_collection_stats'. Before index built, the function returns it is completed.
At that time, your cache will not be able to hold all data, and a lot of time will be spent on loading data.
We will fix the bug recently.

> b After merging segments, would Milvus start to build index immediately?

Yes, it will.

> Will the index-building operation be interrupted by data insertion?

No, i won't.

> c. If I only have one GPU, can I use this GPU to build all 5 collections at the same time? 

No, it can't. Only can build index one by one.

> How about if I have multiple GPUs?

Yes, it can.

> d. For 'preload_collection', after preloading some collections, would these collections keep in Cache stablely[cannot be replaced]?

Not always.
If the cache is full and the data involved in subsequent queries is not in the cache, they will be replaced.

> If the size of real 'preload_collection' is over than 'cache_size', what will happen?

It will stop to load when the cache is full.

> Is there any way for estimating 'preload_collection' size?

[sizing-tool](https://zilliz.com/sizing-tool) may help you.

**Reply by Chisdo** [reply]
Thx for your prompt reply！But I may still have a few confusions about the index. 

1. If I understand correctly, for one partition, [after merge] it should have several index-well-build segments, several segments in the build-index waiting list, and only one segment whose size is less than index_file_size. Am I correct?

2. How can I know the Milvus is building index on the background?

3. After conducting a search, the collections would be loaded into the cache. While if I don't do any operations, how long does the cache would be invalid for the next searching approximately?

4. I also met a problem during rebuilding the index for 2048-dimension vectors. The status error-log shows below. For the 2048 dimension vector, the only 'm' option for it is 64. Hence, may I know how to do the 'useFloat16LokkupTables' trick or how to solve this problem?

![image](https://user-images.githubusercontent.com/36720318/115989155-87db6000-a5ef-11eb-9e62-d27d6db48585.png)

**Reply by yhmo** [reply]
> Thx for your prompt reply！But I may still have a few confusions about the index.
> 
> 1. If I understand correctly, for one partition, [after merge] it should have several index-well-build segments, several segments in the build-index waiting list, and only one segment whose size is less than index_file_size. Am I correct?
> 2. How can I know the Milvus is building index on the background?
> 3. After conducting a search, the collections would be loaded into the cache. While if I don't do any operations, how long does the cache would be invalid for the next searching approximately?
> 4. I also met a problem during rebuilding the index for 2048-dimension vectors. The status error-log shows below. For the 2048 dimension vector, the only 'm' option for it is 64. Hence, may I know how to do the 'useFloat16LokkupTables' trick or how to solve this problem?
> 
> ![image](https://user-images.githubusercontent.com/36720318/115989155-87db6000-a5ef-11eb-9e62-d27d6db48585.png)

1. With continually insert, there could be several index-well-build segments in one partition, and several segments in build index waiting list, but not only one segment whose size is less than index_file_size, because the merge strategy need to consider "write-amplify", please refer this article: https://zhuanlan.zhihu.com/p/148106312  (this article is written in Chinese, you can translate it into English by google translator)

2. Currently we don't have interface to tell client what is going on for the building index on background. For our developers, they can know from log, or meta(MySQL/SQLite).

3. Milvus cache size is defined by server_config.yaml, when you perform a search request, milvus will load required segments from disk into cache(cpu memory). If the cache space is enough, the segments will keep in cache. Once the cache space is fully occupied, milvus will evict some segments that long time not used, according to LRU strategy.

**Reply by shengjun1985** [reply]
> 4 I also met a problem during rebuilding the index for 2048-dimension vectors. The status error-log shows below. For the 2048 dimension vector, the only 'm' option for it is 64. Hence, may I know how to do the 'useFloat16LokkupTables' trick or how to solve this problem?

Please refer to: https://github.com/milvus-io/milvus/issues/3254#issuecomment-690071274
That "(m x 1024) <= MaxSharedMemPerBlock of your graphics card" is a constraint on GPU.
If you want to continue to use GPU, you have to set M smaller.

**Reply by Chisdo** [reply]
> 
> Please refer to: [#3254 (comment)](https://github.com/milvus-io/milvus/issues/3254#issuecomment-690071274)
> That "(m x 1024) <= MaxSharedMemPerBlock of your graphics card" is a constraint on GPU.
> If you want to continue to use GPU, you have to set M smaller.

Since the 'dim/m' at most is 32, for the 2048-dimension vector, the 'm' value at least should be 64 [I cannot decline 'm' value anymore]. On the other hand, the 'm=64' setting is too big for my 2080TI graphic card. So I think the 'useFloat16LokkupTables' trick might be my only solution but I didn't find any interface to configure it. 
 - Is there any way to change 'useFloat16LokkupTables'?
 - Can multiple GPU increase the 'MaxSharedMemPerBlock' of the graphic card?
 - I think there might be a typo here:
 [IVF_PQ](https://milvus.io/docs/v1.0.0/index.md#IVF_PQ)
![image](https://user-images.githubusercontent.com/36720318/116042256-c67d2300-a6a0-11eb-82bf-938d23e79925.png)

**Reply by shengjun1985** [reply]
> Is there any way to change 'useFloat16LokkupTables'?
Can multiple GPU increase the 'MaxSharedMemPerBlock' of the graphic card?

No, it cannot.
This is related to the GPU model but not GPU number.

> I think there might be a typo here.

Thank you very much.
We will fix it soon.

**Reply by Chisdo** [reply]
My pleasure pal. Also appreciate all your answers, which help me a lot!
Btw, I strongly wish to have more parameters to tune for using GPU to build index. It's ok for losing some precision or efficiency. But it's really frustrating for crushing during index building.

**Comment by yhmo** [comment]
Merge operation only happens on segments which size less than index_file_size.
Auto-build-index operation only happens on segment which size larger/equal than index_file_size.

Two ways to create index:
1. create an empty collection, then specify an index type, then insert data until segments merged to exceed index_file_size.
steps:
  create_collection()
  create_index()
  insert()

2. create an empty collection, insert data until size larger than index_file_size, then specify an index type
steps:
  create_collection()
  insert()
  create_index()
  
  
The way_1, milvus automatically build index for segments which size larger than index_file_size
The way_2, milvus will build index for all segments except the tiny segment which row count is less than 4096

For your question ii, I guess you are using the way_1. Since the index type has been specified, you saw all the segments are marked as "PQ", but in fact, their build index tasks maybe are still working, index not finished. So you found the search is brute-force search.

**Reply by Chisdo** [reply]
Hi Yhmo, I get further questions about index building: 
If I create an indexed collection [like SQ8] first, then keep inserting data to the collection [treat it as streaming data]. Is there any possibility that I can drop and rebuild the index of the collection while inserting data?

**Reply by yhmo** [reply]
Milvus has three queues to process different requests:
- create/insert/delete requests put into "DDL" queue
- search requests put into "DQL" queue
- others requests put into another queue

Consider this case:
- client_1 continually call insert() to insert vectors.
- client_2 call drop_index() and create_index()

When client_2 call create_index(), the "DDL" queue will blocked until all segments finish index. At the same time, the client_1 will paused since the queue is blocked by create_index().
But if the client_2 stopped, by ctrl+c or connection break, etc, the create_index request will be removed from the "DDL" queue, and put to background thread to execute, therefore the client_1 can continue insert.
